### ISSUE TITLE
Splash screen at startup

### ISSUE THREAD

**Issue by Pegoth** [rmcrackan/Libation#322]
I noticed that the software starts a bit slow, which is not a problem, but is it possible to have a splash screen appear instantly with "Loading" text or something? Currently it feels like it failed to start, only to open ~20-30 sec later.

**Comment by rmcrackan** [comment]
Thanks for the suggestion. That's a good idea.

**Comment by Mbucari** [comment]
@rmcrackan we need to solve the problems with the publish binaries. The publish builds are the only way that you can make ready to run binaries, which will dramatically speed up startup times.

**Comment by rmcrackan** [comment]
I had a similar thought. A splash screen is a good idea but it's also a band aid; the real solution is to figure out why it takes so long to load. Part of this will be sped up when figure out the 'ready to run' publish. The other part is: it hasn't always taken this long -- what in the world has nickel and dimed us to this place?

Off the top of my head: maybe there's some of the pre-load work can run as parallel instead of serial, and we can use `WhenAll` to block until they're all done. This kind of profiling and optimizing will be its own project and we have enough going on right now.

**Comment by Mbucari** [comment]
> maybe there's some of the pre-load work can run as parallel instead of serial

They're definitely is, and I spent a lot of time doing just that for the avalonia build. One of the big things that shaved off of decent amount of time was to begin loading the library in parallel asap, then pass the list to the main form to load into the grid.

Another one would be checking for updates, which should definitely happen in parallel.  Remember way back when when you said it should be a seamless process? Just have it run as a background task and download the zip into a temp location, then add the update file's path to appsettings.json. The next time libation is started, it will see that it has an update package downloaded and available and will run the installer on it.

Also, I noticed a couple of Thread sleeps in Configuration that I am pretty sure are unnecessary. They come after writing a file, but they should not be necessary because after the write method returns, the file should already be available for reading.

**Comment by rmcrackan** [comment]
> Another one would be checking for updates, which should definitely happen in parallel.

Great point. We can continue that discussion in the PR comments where we're already discussing autoupdater.

> Remember way back when when you said it should be a seamless process? ...

Yeah. That's probably the right answer, especially since we're having to re-think auto update anyway.

> should not be necessary because after the write method returns, the file should already be available for reading.

You'd think that until you get burned by it. There's actually a slight delay between when windows returns and when it's actually ready, especially after initial file creation (as opposed to edit). My typically 100ms is probably too aggressive. We could halve the sleep after file creation and maybe even eliminate it if we're positive the file already exists.

Also in my to-do discussion list for later today will be ready to run publish questions. So we can handle those then. And that will contribute to speed up efforts.

That said: do we want to nix a splash screen in favor of this other stuff?

**Comment by Mbucari** [comment]
> You'd think that until you get burned by it.

Wow. I haven't been burned by it, but that's a terrible system right there.

As I was trying to wrap my head around just what was happening in Configuration, it occurred to me that it could be made much simpler if you interacted with appsettings.json through the json persister instead of reading/writing the the file text.  Maybe that would also solve the delay problem?

> That said: do we want to nix a splash screen in favor of this other stuff?

Startup times on Linux pub builds are 2-4 seconds. That should be the benchmark. Is that slow enough for you to justify a splash screen?

**Comment by rmcrackan** [comment]
> Configuration

There are a lot of old choices in there which haven't been re-thought in a loooong time. I have no doubt there is room for optimization. That has to be done though like you're handling nitroglycerin. If anything goes wrong with configuration or appsettings you can effortlessly brick the app. (I should know -- I've done it several times over the years. It is not a fun conversation to have with users.)

> 2-4 seconds

Under 2 seconds would be ideal but probably unrealistic. If it's under 6 seconds though, I don't see the need for a splash screen.

Since the soon to be released version will significantly speed up load time, I'm going to close this issue. If we find that it doesn't speed it up enough, we can re-open it at that time.

**Comment by rmcrackan** [comment]
> Wow. I haven't been burned by it, but that's a terrible system right there.

Either windows or .net have finally fixed this and what I said is no longer true. I've been burned by it MANY times in the past. Perhaps it was on .net framework and not core. In any case, I just ran a bunch of tests and I could not get this bug to happen. It looks like you are correct and these sleeps around file writing are no longer needed at all. Thanks for making me re-evaluate this.

**Comment by Mbucari** [comment]
Another thing to consider is that with Avalonia, getting to the point where you could display a splash screen (UI libs are loaded and ready) accounts for most of the startup time. So there's really no benefit in the Avalonia builds.
